fluidplayer.js

```diff
@@ -697,18 +697,65 @@ var fluidPlayerClass = {
         var player = this;
         var videoPlayerTag = document.getElementById(this.videoPlayerId);
         var list = [];
-        list.length = 0;
+        var vastList = Object.values(player.adList);
+        
+        if (roll == 'preRoll') {
+            for (var [key, vast] of vastList.entries()) {
+                if (vast.roll == 'preRoll' && vast.played == false) {
+                    var firstPreRoll = vast.id;
+                    break;
+                }
+            }
 
-        list = player.findRoll(roll);
+            player.processUrl(player.adList[firstPreRoll].vastTag, firstPreRoll);
+            videoPlayerTag.addEventListener('adId_' + firstPreRoll, player[roll]);
+        } else if (roll === 'midRoll') {
+            var midRolls = [];
 
-        for (var i = 0; i < list.length; i++) {
-            var adListId = list[i];
+            for (var [key, vast] of vastList.entries()) {
+                for (var [secKey, secVast] of vastList.entries()) {
```

Comment:
Do not like these multiple iterations over all vastList entries with checks like vastList[key].roll === 'midRoll'.
They are a total waste and they are repeated below multiple times here and in onVastAdEnded as well to do something like midRolls.push(vastList[secKey]);

Why don't we restructure the vastList to have roll type as key to begin with? Or run one full cycle and build ONCE the object with such keys and have it saved?

Another thing: all the vast tags used to be loaded and parsed at once. Now it's just first Vast tag for each type preRoll, midRoll, postRoll at the beginning. And then the next ones are processed "As we go" in onVastAdEnded event.
I like the whole "As we go" principle but the current code has become very convoluted to follow.
I think with better structure we could push them all into correct object. And then process as we go from it completely (in one place). Starting from first pre roll and up to whatever comes next. Without doing these things in two different places.
This may take a bit of refactoring of things like vastList, timerPool and adPool, but would really be beneficial.